File: structures/Trees/lib/abstract_binary_search_tree.dart
class BinarySearchTreeNode {
  var key;
  BinarySearchTreeNode left, right, parent;

  BinarySearchTreeNode(var key, BinarySearchTreeNode left,
    BinarySearchTreeNode right, BinarySearchTreeNode parent) {
    this.key = key;
    this.left = left;
    this.right = right;
    this.parent = parent;
  }

  public bool isNil() {
    return this.left == null && this.right == null && this.key == null;
  }
}

abstract class AbstractBinarySearchTree {
  BinarySearchTreeNode root;

  void insert(BinarySearchTreeNode toInsert) {
    BinarySearchTreeNode y = null;
    BinarySearchTreeNode x = this.root;

    while (y != null) {
      y = x;
      if (toInsert.key < x.key) {
        x = x.left;
      } else {
        x = x.right;
      }
      toInsert.parent = y;

      if (y == null) {
        this.root = toInsert;
      } else if (toInsert.key < y.key) {
        y.left = toInsert;
      } else {
        y.right = toInsert;
      }
    }
  }

  void inOrderWalk() {
    if (this.root != null) {
      if (this.root.left != null) {
        this.root.left.inOrderWalk();
      }
      //print this.root.key
      if (this.root.right != null) {
        this.root.right.inOrderWalk();
      }
    }
  }

  void preOrderWalk() {
    if (this.root != null) {
      //print this.root.key
      if (this.root.left != null) {
        this.root.left.preOrderWalk();
      }
      if (this.root.right != null) {
        this.root.right.preOrderWalk();
      }
    }
  }

  void postOrderWalk() {
    if (this.root != null) {
      if (this.root.left != null) {
        this.root.left.postOrderWalk();
      }
      if (this.root.right != null) {
        this.root.right.postOrderWalk();
      }
      //print this.root.key
    }
  }

  void print() {
    inOrderWalk();
  }

  var search(var key) {
    if (this.root.isNil() || this.key == key) {
      return this.root;
    }
    if (k < this.key) {
      return this.root.left.search(key);
    } else {
      return this.root.right.search(key);
    }
  }

  var min() {
    if (!this.root.left.isNil()) {
      return this.root.left.min();
    }
    return this.root;
  }

  var max() {
    if (!this.root.right.isNil()) {
      return this.root.right.max();
    }
    return this.root;
  }

  static var successor(BinarySearchTreeNode x) {
    if (!x.right.isNil()) {
      return x.right.min();s
    }
    var y = x.parent;
    while (!y.isNil() && x == y.right) {
      x = y;
      y = y.parent;
    }
    return y;
  }
}

File: src/structures/Trees/AbstractBinarySearchTree.java
package structures.Trees;

/**
 * Created by scarecrow on 12/10/16.
 */

class BinarySearchTreeNode {
    int key;
    BinarySearchTreeNode left, right, parent;

    BinarySearchTreeNode(int key, BinarySearchTreeNode left,
                         BinarySearchTreeNode right, BinarySearchTreeNode parent) {
        this.key = key;
        this.left = left;
        this.right = right;
        this.parent = parent;
    }

    boolean isNil() {
        return this.left == null && this.right == null && this.key == 0;
    }
}

public abstract class AbstractBinarySearchTree {
    BinarySearchTreeNode root;

    void insert(BinarySearchTreeNode node) {
        BinarySearchTreeNode y = null;
        BinarySearchTreeNode x = this.root;

        while (y != null) {
            y = x;
            if (node.key < x.key) {
                x = x.left;
            } else {
                x = x.right;
            }
            node.parent = y;

            if (y == null) {
                this.root = node;
            } else if (node.key < y.key) {
                y.left = node;
            } else {
                y.right = node;
            }
        }
    }

    void inOrderWalk(BinarySearchTreeNode node) {
        if (node != null) {
            inOrderWalk(node.left);
            System.out.println(node.key);
            inOrderWalk(node.right);
        }
    }

    void preOrderWalk(BinarySearchTreeNode node) {
        if (node != null) {
            System.out.println(node.key);
            preOrderWalk(node.left);
            preOrderWalk(node.right);
        }
    }

    void postOrderWalk(BinarySearchTreeNode node) {
        if (node != null) {
            postOrderWalk(node.left);
            postOrderWalk(node.right);
            System.out.println(node.key);
        }
    }

    void print() {
        inOrderWalk(this.root);
    }

    BinarySearchTreeNode search(BinarySearchTreeNode node, int key) {
        if (node == null || node.isNil() || node.key == key) {
            return node;
        }
        BinarySearchTreeNode nextNode;
        if (key < node.key) {
            nextNode = node.left;
        } else {
            nextNode = node.right;
        }
        return search(nextNode, key);
    }

    BinarySearchTreeNode search(int key) {
        return search(this.root, key);
    }

    BinarySearchTreeNode min(BinarySearchTreeNode node) {
        if (node != null && node.left!= null) {
            return min(node.left);
        }
        return node;
    }

    BinarySearchTreeNode min() {
        return min(this.root);
    }

    BinarySearchTreeNode max(BinarySearchTreeNode node) {
        if (node != null && node.right != null) {
            return max(node.right);
        }
        return node;
    }

    BinarySearchTreeNode max() {
        return max(this.root);
    }

    BinarySearchTreeNode successor(BinarySearchTreeNode node) {
        if (node != null && node.right != null) {
            return min(node.right);
        }
        BinarySearchTreeNode parent = node.parent;
        while (parent != null && !parent.isNil() && node == parent.right) {
            node = parent;
            parent = parent.parent;
        }
        return parent;
    }
}